### DISCUSSION TITLE
How to get accurate diagnostics when viewing files included from external libraries?

### DISCUSSION THREAD

**Discussion by recursivenomad** [clangd/clangd#2075] [Q&A] [answered]
I am developing a project in VSCodium with the clangd extension that looks like this:

```
|
|__/project
|      |
|      |__/build
|      |     |
|      |     |__/compile_commands.json
|      |
|      |__/main.c
|      |
|      |__/defs.h
|
|
|__/sdk
      |
      |__/lib.h
      |
      |__/lib.c
```

Everything with the language server looks fine when I am viewing `main.c`, and definitions from `lib.h` are reflected accurately in `main.c`.

When I follow the `#include` to `lib.h` however, the definitions made in `defs.h` and compile_commands.json are not recognized by the language server.

To get around this issue, I tried making a symbolic link to the sdk, setting up CMake to reference the link, like this:

```
|
|__/project
|      |
|      |__/build
|      |     |
|      |     |__/compile_commands.json
|      |
|      |__/lib
|      |     |
|      |     |__/sdk (symlink) ->
|      |
|      |__/main.c
|      |
|      |__/defs.h
|
|
|__/sdk
      |
      |__/lib.h
      |
      |__/lib.c
```

This allowed me to manually navigate into the symlinked files within the VSCode UI, and this manual navigation *does* successfully render changes to `lib.h` definitions made in `defs.h` and compile_commands.json.  However, if I follow the `#include "lib.h"` found in `main.c`, it still takes me to the not-symbolic path, and the diagnostic link is still broken.

The third thing I tried was to straight up copy the whole SDK into my project:

```

project
   |
   |__/build
   |     |
   |     |__/compile_commands.json
   |
   |__/lib
   |     |
   |     |__/sdk
   |
   |__/main.c
   |
   |__/defs.h

```

And... yeah, this worked; but that kind of defeats the purpose of being able to link to an sdk/library 😂

So, seeing as using external libraries is an incredibly common practice, is there something I can do differently which will allow me to see full, accurate, project-specific diagnostic information of the sdk (just like when it's directly in my project), but keeping it in a central, external location?

Linking to a few issues that appear to relate:
- #197 
- #544 
- #1929

**Comment by HighCommander4** [comment]
Have a look at https://clangd.llvm.org/faq#how-do-i-fix-errors-i-get-when-opening-headers-outside-of-my-project-directory

**Reply by recursivenomad** [reply]
This appears to work exactly how I need it to, thank you!  I thought I'd already tried this, but perhaps I still had symlinks causing problems when I first attempted.

And what's even better is that you can give `--compile-commands-dir=` relative paths so I won't have to set up my project manually each time!!  Now my experience with this language server has gone from helpful to downright delightful - Wow 😎

I am curious, why doesn't clangd simply hold onto the project's automatically-discovered compile_commands.json by default?  It seems silly from my narrow perspective to need to manually specify the location in order for external analysis to work properly.

**Reply by HighCommander4** [reply]
> why doesn't clangd simply hold onto the project's automatically-discovered compile_commands.json by default?

The short answer is that clangd doesn't really have a concept of a "project". (The reason for that in turn is that it has to support a variety of clients, not all of which themselves have a concept of a project or workspace.)

The state that clangd keeps is either: 

 1. Per clangd invocation, based on information provided at the time clangd is started, e.g. through command-line arguments, such as `--compile-commands-dir`.
 2. Per file. This is how `compile_commands.json` discovery works without `--compile-commands-dir`: every time a file is opened, clangd searches the ancestor directories of the file (and their `build` siblings) for `compile_commands.json`.

One way to think of it is: if you're working with an editor which has the concept of a workspace, then `--compile-commands-dir` is the way of telling clangd what the workspace directory is.

**Reply by recursivenomad** [reply]
Thanks for the explanation, that all makes a lot of sense :)

With the clangd VSCode extension, I can add a list of clangd launch commands.  VSCode allows us to add such settings so on a per-project scale OR a global scale (but they don't overlap - a project setting overwrites a global setting).  So my predicament now is that globally I have `--query-driver` pointing to my compilers, and was hoping to set `--compile-commands-dir` on a per-project scale...

Am I able to offload `--query-driver` to my user config somehow, thus freeing up my VSCode clangd launch arguments for a per-project `--compile-commands-dir` argument?  Or is there a way to provide a list to `--compile-commands-dir` to search through before defaulting to per-file behaviour?

**Reply by HighCommander4** [reply, marked as answer]
> With the clangd VSCode extension, I can add a list of clangd launch commands. VSCode allows us to add such settings so on a per-project scale OR a global scale (but they don't overlap - a project setting overwrites a global setting). So my predicament now is that globally I have `--query-driver` pointing to my compilers, and was hoping to set `--compile-commands-dir` on a per-project scale...
> 
> Am I able to offload `--query-driver` to my user config somehow, thus freeing up my VSCode clangd launch arguments for a per-project `--compile-commands-dir` argument? Or is there a way to provide a list to `--compile-commands-dir` to search through before defaulting to per-file behaviour?

Interesting question!

After poking around a bit, I discovered the following solution. In VSCode's global settings, add:

```json
"clangd.arguments": [
  "--query-driver=/path/to/your/compiler",
  "--compile-commands-dir=${workspaceFolder}"
]
```

The `${workspaceFolder}` gets expanded to the full path of the workspace directory in each workspace (project), saving you the trouble of having to define it separately in each one.

**Reply by recursivenomad** [reply]
Ahh, this one is so close!  It looks like I still need to manually specify the location of my build directory (or its project directory one-level-up).  The reason this remains a problem for me is that for some projects my build directory is in `project-root/build`, while others it is in `project-root/firmware/build`, etc 🙈  Adding more than one `--compile-commands-dir` to the list appears to only apply the last one.

I tried exploring the curiously relevant `"clangd.fallbackFlags"` using:

```json
"clangd.fallbackFlags": [
  "--compile-commands-dir=${workspaceFolder}/firmware/build"
]
```

in my per-project `settings.json`, and with `--log=verbose` it *does* correctly note the presence of the project-specific compile_commands.json - but unfortunately it doesn't get applied 😕  I'm not comprehending everything in the verbose outputs, but searching for my path to compile_commands.json, this stood out to me (see highlight):

> V >>> {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"drv_unsupported_opt","message":<mark>"Unsupported option '--compile-commands-dir=d:/path/to/firmware/build'"</mark>,"range":{"end":{"character":0,"line":0},"start":{"character":0,"line":0}},"relatedInformation":[],"severity":1,"source":"clang"}, `...`

**Reply by HighCommander4** [reply]
> The reason this remains a problem for me is that for some projects my build directory is in `project-root/build`, while others it is in `project-root/firmware/build`, etc 🙈

One common pattern I've seen is symlinking the `compile_commands.json` from the build directory to the project root.

> I tried exploring the curiously relevant `"clangd.fallbackFlags"` using:
> 
> ```json
> "clangd.fallbackFlags": [
>   "--compile-commands-dir=${workspaceFolder}/firmware/build"
> ]
> ```
> 
> in my per-project `settings.json`, and with `--log=verbose` it _does_ correctly note the presence of the project-specific compile_commands.json - but unfortunately it doesn't get applied 😕 I'm not comprehending everything in the verbose outputs, but searching for my path to compile_commands.json, this stood out to me (see highlight):
> 
> > V >>> {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"code":"drv_unsupported_opt","message":"Unsupported option '--compile-commands-dir=d:/path/to/firmware/build'","range":{"end":{"character":0,"line":0},"start":{"character":0,"line":0}},"relatedInformation":[],"severity":1,"source":"clang"}, `...`

`clangd.fallbackFlags` are flags to put into compile commands (like `-I` include directories, `-D` defines, etc. -- stuff that would go into the entries of `compile_commands.json`), rather than command-line flags to clangd itself (like `--compile-commands-dir`).

**Reply by recursivenomad** [reply]
Ahh, I see - alright, I will go ahead and try that for now; but in the meantime, I think I will open [an issue in the VSCode extension](https://github.com/clangd/vscode-clangd/issues/644) to consider adding a separate option for the value of `--compile-commands-dir`, so that it can be overwridden on a per-project level 🙂

Thanks again for all of your thorough help!

**Reply by recursivenomad** [reply]
For anyone in a similar situation, if you are using the [CMake Tools](https://github.com/microsoft/vscode-cmake-tools) VSCode extension, it appears you can also specify a location to automatically copy your compile_commands.json to.

So now my global `settings.json` looks something like:

```json
"clangd.arguments": [
  "--query-driver=path/to/compiler/bin/*",
  "--query-driver=path/to/another/compiler/bin/*",
  "--compile-commands-dir=${workspaceFolder}"
],
"cmake.copyCompileCommands": "${workspaceFolder}/compile_commands.json"
```

*(Note that as of writing, [there is a bug on Windows](https://github.com/microsoft/vscode-cmake-tools/issues/3214#issuecomment-2198199560) where it doesn't protect from colliding with the original compile_commands.json location, so I recommend copying to the project root, rather than `project-root/build` until that issue is resolved)*
